sql-plugin/src/main/scala/com/nvidia/spark/rapids/profiler.scala

```diff
@@ -0,0 +1,387 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import java.nio.ByteBuffer
+import java.nio.channels.{Channels, WritableByteChannel}
+import java.util.concurrent.{ConcurrentHashMap, Future, ScheduledExecutorService, TimeUnit}
+import java.util.regex.Pattern
+
+import scala.collection.mutable
+
+import com.nvidia.spark.rapids.jni.Profiler
+import org.apache.hadoop.fs.Path
+import org.apache.hadoop.ipc.CallerContext
+
+import org.apache.spark.{SparkContext, TaskContext}
+import org.apache.spark.api.plugin.PluginContext
+import org.apache.spark.internal.Logging
+import org.apache.spark.io.CompressionCodec
+import org.apache.spark.scheduler.{SparkListener, SparkListenerJobEnd, SparkListenerStageCompleted}
+import org.apache.spark.sql.rapids.execution.TrampolineUtil
+import org.apache.spark.util.SerializableConfiguration
+
+object ProfilerOnExecutor extends Logging {
+  private val jobPattern = Pattern.compile("SPARK_.*_JId_([0-9]+).*")
+  private var writer: Option[ProfileWriter] = None
+  private var timeRanges: Option[Seq[(Long, Long)]] = None
+  private var jobRanges: RangeConfMatcher = null
+  private var stageRanges: RangeConfMatcher = null
+  // NOTE: Active sets are updated asynchronously, synchronize on ProfilerOnExecutor to access
+  private val activeJobs = mutable.HashSet[Int]()
+  private val activeStages = mutable.HashSet[Int]()
+  private var timer: Option[ScheduledExecutorService] = None
+  private var timerFuture: Option[Future[_]] = None
+  private var driverPollMillis = 0
+  private val startTimestamp = System.nanoTime()
+  private var isProfileActive = false
+
+  def init(pluginCtx: PluginContext, conf: RapidsConf): Unit = {
+    require(writer.isEmpty, "Already initialized")
```

Comment:
Do we want lots of warning if profiling is enabled? I am a bit concerned about someone leaving this on in production on accident.